src/event_stream.rs

```diff
@@ -0,0 +1,62 @@
+use std::sync::mpsc::TryRecvError;
+
+/// Trait to fake a channel.
+pub trait EventStream {
+    /// Item produced by this stream.
+    type Item;
+    /// Error type produced if something goes wrong or no items are available.
+    type Error;
+
+    /// Attempt to fetch the next item from the stream.
+    ///
+    /// Implementations should be non-blocking and return an appropriate error if no items
+    /// are available.
+    fn try_next_ev(&mut self) -> Result<Self::Item, Self::Error>;
+
+    /// Process events, storing them on the internal buffer.
+    ///
+    /// After calling poll, any events produced will be accessible via try_next_ev.
+    fn poll(&mut self) -> bool;
+}
+
+/// Trait for state machines and other event producers who produce multiple events at once.
+pub trait EventStepper {
+    type Item;
+
+    /// Produce multiple events and return them, or Err(()) if something goes wrong.
+    /// TODO: propagate actual errors better.
```

Comment:
This might be a good idea ;) Looks like a big task, though, so maybe it will be better to leave it for another PR.